### ISSUE TITLE
JSONDecodeError

### ISSUE THREAD

**Issue by ewaizel** [CiscoDevNet/DNAC-TemplateProgrammer#4]
I'm trying to run this using my own DNAC server (version 3.2.2.4). Any idea?

File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1264.0_x64__qbz5n2kfra8p0\lib\json\__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1264.0_x64__qbz5n2kfra8p0\lib\json\decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1264.0_x64__qbz5n2kfra8p0\lib\json\decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None

**Comment by aradford123** [comment]
could you share how you ran the script please?

as well as the full traceback?

**Comment by ewaizel** [comment]
Thank you for your quick reply.

This is what I’m using:

deploy.py --template "LNC Lab/CBS Default port range" --device 10.45.0.4 --params '{"Last":"14","First":"11"}'

The template name was identified after running the script with no parameters.
The associated device is already part of the inventory.

The initial template I created was very simple. It was done in velocity (VTL) and was expecting to receive the first and last ports (parameters = $First and $Last).
You can see the details of the “CBS Default port range” template below.

Hope this helps and thanks again.

Enrique Waizel


## Catalyst 9300 port range script
#set ($Version = "0.1")
## This script version
## Date Apr 4, 2022
!Created using port range script verion $Version

## $First
##set ($First = "10")
!First = $First
## $Last
##set ($Last = "15")
!Last = $Last

int range g1/0/$First - $Last
description default


This is the output during the execution:


E:\ew\OneDrive\Documents\Proyectos\CBS\Doc\Switching and Routing\L2 L3 replacement\DNAC>deploy.py --template "LNC Lab/CBS Default port range" --device 10.45.0.4 --params '{"Last":"14","First":"11"}'
Looking for: LNC Lab/CBS Default port range
https://10.29.0.47:443/dna/intent/api/v1/template-programmer/template
TemplateId: a5a3f1b9-4214-4d8f-8135-65aa30010033 Version: 5

https://10.29.0.47:443/dna/intent/api/v1/template-programmer/template/a5a3f1b9-4214-4d8f-8135-65aa30010033
Showing Template Body:
{'name': 'CBS Default port range', 'description': 'Apply the same for a range of interfaces', 'author': 'admin', 'deviceTypes': [{'productFamily': 'Switches and Hubs', 'productSeries': 'Cisco Catalyst 9300 Series Switches'}], 'softwareType': 'IOS-XE', 'softwareVariant': 'XE', 'templateContent': '## CBS Catalyst 9300 port range script\n#set ($Version = "0.1")\n## This script version\n## Date Apr 4, 2022\n!Created using port range script verion $Version\n\n## $First \n##set ($First = "10")\n!First = $First\n## $Last \n##set ($Last = "15")\n!Last = $Last\n\nint range g1/0/$First - $Last \ndescription default\n\n\n', 'rollbackTemplateContent': '', 'templateParams': [{'parameterName': 'First', 'dataType': 'STRING', 'defaultValue': None, 'description': None, 'required': True, 'notParam': False, 'paramArray': False, 'displayName': None, 'instructionText': None, 'group': None, 'order': 1, 'customOrder': 0, 'selection': None, 'range': [], 'key': None, 'provider': None, 'binding': '', 'id': '2f656725-e066-4f25-af56-a2d7ea7d2d39'}, {'parameterName': 'Last', 'dataType': 'STRING', 'defaultValue': None, 'description': None, 'required': True, 'notParam': False, 'paramArray': False, 'displayName': None, 'instructionText': None, 'group': None, 'order': 2, 'customOrder': 0, 'selection': None, 'range': [], 'key': None, 'provider': None, 'binding': '', 'id': '3fedb064-ee48-4d5f-befd-00a042cda78e'}], 'rollbackTemplateParams': [], 'composite': False, 'containingTemplates': [], 'language': 'VELOCITY', 'id': 'a5a3f1b9-4214-4d8f-8135-65aa30010033', 'version': '5', 'customParamsOrder': False, 'createTime': 1647646863562, 'lastUpdateTime': 1649437003066, 'latestVersionTime': 1649437020054, 'projectName': 'LNC Lab', 'projectId': '03944ff6-a014-4343-ad5c-87daed7766ba', 'parentTemplateId': 'a26425d2-1f1e-4569-bd35-50913dc99c5f', 'validationErrors': {'templateErrors': [], 'rollbackTemplateErrors': [], 'templateId': 'a5a3f1b9-4214-4d8f-8135-65aa30010033', 'templateVersion': None}, 'projectAssociated': True, 'documentDatabase': False}
## CBS Catalyst 9300 port range script
#set ($Version = "0.1")
## This script version
## Date Apr 4, 2022
!Created using port range script verion $Version

## $First
##set ($First = "10")
!First = $First
## $Last
##set ($Last = "15")
!Last = $Last

int range g1/0/$First - $Last
description default




Required Parameters for template body: {"First":"","Last":""}

Bindings []

Executing template on:10.45.0.4, with Params:'{Last:14,First:11}'
Traceback (most recent call last):
  File "E:\ew\OneDrive\Documents\Proyectos\CBS\Doc\Switching and Routing\L2 L3 replacement\DNAC\deploy_sample_2.py", line 244, in <module>
    response = execute(id, params, bindings, args.device, inputParams, args.force)
  File "E:\ew\OneDrive\Documents\Proyectos\CBS\Doc\Switching and Routing\L2 L3 replacement\DNAC\deploy_sample_2.py", line 62, in execute
    "params": json.loads(params)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1264.0_x64__qbz5n2kfra8p0\lib\json\__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1264.0_x64__qbz5n2kfra8p0\lib\json\decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1264.0_x64__qbz5n2kfra8p0\lib\json\decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)

**Comment by aradford123** [comment]
It has not got to the stage of pushing the template.

This is strange.. it is unhappy with the json you have provided for the params

>File "E:\ew\OneDrive\Documents\Proyectos\CBS\Doc\Switching and Routing\L2 L3 replacement\DNAC\deploy_sample_2.py", line 62, in execute 
"params": json.loads(params) 

That looks like valid JSON --params '{"Last":"14","First":"11"}'.  unless it is something  strange with the way windows handles quotes, and nested quotes?

**Comment by ewaizel** [comment]
I found the issue. I'm running this in Windows. I had to use some escape characters to do the trick. Also, single quotation marks are not needed.
deploy.py --template "LNC Lab/CBS Default port range" --device 10.45.0.4 --params {\"Last\":\"13\",\"First\":\"11\"} --force

The bulk csv deployment worked fine:
deploy.py --template "LNC Lab/CBS Default port range" --bulkfile testcsv.csv --force

Thank you.
